src/commissaire_service/investigator/__init__.py

```diff
@@ -0,0 +1,204 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import datetime
+import json
+
+import commissaire.constants as C
+
+from commissaire.models import Cluster, Host, Network
+from commissaire.util.ssh import TemporarySSHKey
+from commissaire.util.config import ConfigurationError
+
+from commissaire_service.oscmd import get_oscmd
+from commissaire_service.service import CommissaireService
+from commissaire_service.transport import ansibleapi
+
+
+class InvestigatorService(CommissaireService):
+    """
+    Investigates new hosts to retrieve and store facts.
+    """
+
+    def __init__(self, exchange_name, connection_url):
+        """
+        Creates a new InvestigatorService.
+
+        :param exchange_name: Name of the topic exchange
+        :type exchange_name: str
+        :param connection_url: Kombu connection URL
+        :type connection_url: str
+        """
+        queue_kwargs = [
+            {'routing_key': 'jobs.investigate'}
+        ]
+        CommissaireService.__init__(
+            self, exchange_name, connection_url, queue_kwargs)
+
+    def _get_etcd_config(self):
+        """
+        Extracts etcd configuration from a registered store handler.
+        If no matching handler is found, return defaults for required
+        values.
+
+        :returns: A dictionary of configuration values
+        :rtype: dict
+        """
+        response = self.request('storage.list_store_handlers')
+        for handler in response.get('result', []):
+            if handler['handler_type'] == 'EtcdStoreHandler':
+                return handler['config']
+
+        raise ConfigurationError(
+            'Configuration is missing an EtcdStoreHandler')
+
+    def _get_cluster_and_network_models(self, cluster_data):
+        """
+        Creates cluster and network models from the given cluster data.
+
+        :param cluster_data: Data for a cluster
+        :type cluster_data: dict
+        :returns: a Cluster and Network model
+        :rtype: tuple
+        """
+        try:
+            cluster = Cluster.new(**cluster_data)
+            network = Network.new(name=cluster.network)
+            params = {
+                'model_type_name': network.__class__.__name__,
+                'model_json_data': network.to_json()
+            }
+            response = self.request('storage.get', params=params)
+            network = Network.new(**response['result'])
+        except TypeError:
+            cluster = Cluster.new(type=C.CLUSTER_TYPE_HOST)
```

Comment:
I added logging for these results, but put it just before the `transport.bootstrap()` call.